HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc

```diff
@@ -0,0 +1,235 @@
+#include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalStage1TruncationImpl_SA.h"
+#include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalStage1TruncationConfig_SA.h"
+
+#include "HGCTPGStage1Emulator/interface/SAConfigParser.h"
+#include "HGCTPGStage1Emulator/interface/SATCParser.h"
+
+#include <iostream>
+#include <fstream>
+#include <algorithm>
+#include <cmath>
+#include <memory>
+
+using namespace l1thgcfirmware;
+
+typedef std::vector < HGCalTriggerCell > HGCalTriggerCells;
+typedef std::vector < HGCalTriggerCells > HGCalTriggerCellsPerBx;
+
+class HGCTPGEmulatorTester {
+
+public:
+
+  HGCTPGEmulatorTester(const std::string, const std::string);
+  ~HGCTPGEmulatorTester(){};
+  
+  int runAlgo(const HGCalTriggerCellsPerBx&, HGCalTriggerCellsPerBx&);
+  int dumpTCs(const HGCalTriggerCellsPerBx&, const std::string);
+  int dumpTCs_fw(const HGCalTriggerCellsPerBx&, const std::string);
+
+  //getter
+  HGCalTriggerCellsPerBx getTCs() {return allTCs_;}
+
+private:
+  // dummy initialization of configuration; overwritten later
+  Stage1TruncationConfig stage1Config_ = Stage1TruncationConfig(false,
+								0.,
+								0.,
+								0,
+								std::vector<unsigned>{0},
+								std::vector<double>{0.});
+
+  double rzmin_;
+  double rzmax_;
+  double rz_bin_size_;
+  static constexpr double margin_ = 1.001;
+
+  TCMap theTCMap_;
+  HGCalStage1TruncationImplSA stage1Algo_;
+
+  HGCalTriggerCellsPerBx allTCs_;
+
+  static constexpr unsigned offset_roz_ = 1;
+  static constexpr unsigned mask_roz_ = 0x3f;  // 6 bits, max 64 bins
+  static constexpr unsigned mask_phi_ = 1;
+
+  static constexpr unsigned offset_e_ = 8; // energy on 8 bits (0->255)
+  
+};
+
+HGCTPGEmulatorTester::HGCTPGEmulatorTester(const std::string jsonConfigFile, 
+					   const std::string tcList)
+{
+  
+  
+  std::cout << "-- Initializing config reader" << std::endl;
+  SAConfigParser cfgReader(jsonConfigFile);
+  stage1Config_ = cfgReader.getCfg();
+
+  rzmin_ = stage1Config_.rozMin();
+  rzmax_ = stage1Config_.rozMax();
+  rz_bin_size_ = (rzmax_ - rzmin_) * margin_ / stage1Config_.rozBins();
+
+  theTCMap_ = cfgReader.getTCmap();
+
+  std::cout << "-- Initializing TC reader" << std::endl;
+  SATCParser tcReader(tcList, theTCMap_);
+  std::cout << "-- Getting TC list" << std::endl;
+  allTCs_ = tcReader.TClist();
+
+}
+
+int HGCTPGEmulatorTester::runAlgo(const HGCalTriggerCellsPerBx& tcPerBx, HGCalTriggerCellsPerBx& sortedTruncatedTCsPerBx) {
+  
+  unsigned int counter = 0;
+  for (const auto TCs : tcPerBx) {
+    // Run the algorithm
+    HGCalTriggerCells sortedTruncatedTCs;
+    unsigned error_code = stage1Algo_.run(TCs, stage1Config_, sortedTruncatedTCs);
+    sortedTruncatedTCsPerBx.push_back(sortedTruncatedTCs);
```

Comment:
Couldn't the `sortedTruncatedTCsPerBx` be directly filled instead of copying things around?